test/Microsoft.AspNetCore.Razor.Language.Test/CodeGeneration/PreallocatedAttributeTargetExtensionTest.cs

```diff
@@ -0,0 +1,207 @@
+            {
+                Writer = new CSharpCodeWriter()
+            };
+
+            var node = new AddPreallocatedTagHelperHtmlAttributeIRNode()
+            {
+                VariableName = "_tagHelper1"
+            };
+
+            // Act
+            extension.WriteAddPreallocatedTagHelperHtmlAttribute(context, node);
+
+            // Assert
+            var csharp = context.Writer.Builder.ToString();
+            Assert.Equal(
+@"__tagHelperExecutionContext.AddHtmlAttribute(_tagHelper1);
+",
+                csharp,
+                ignoreLineEndingDifferences: true);
+        }
+
+        [Fact]
+        public void WriteDeclarePreallocatedTagHelperAttribute_RendersCorrectly()
+        {
+            // Arrange
+            var extension = new PreallocatedAttributeTargetExtension();
+            var context = new CSharpRenderingContext()
+            {
+                Writer = new CSharpCodeWriter()
+            };
+
+            var node = new DeclarePreallocatedTagHelperAttributeIRNode()
+            {
+                Name = "Foo",
+                Value = "Bar",
+                ValueStyle = HtmlAttributeValueStyle.DoubleQuotes,
+                VariableName = "_tagHelper1",
+            };
+
+            // Act
+            extension.WriteDeclarePreallocatedTagHelperAttribute(context, node);
+
+            // Assert
+            var csharp = context.Writer.Builder.ToString();
+            Assert.Equal(
+@"private static readonly global::Microsoft.AspNetCore.Razor.TagHelpers.TagHelperAttribute _tagHelper1 = new global::Microsoft.AspNetCore.Razor.TagHelpers.TagHelperAttribute(""Foo"", ""Bar"", global::Microsoft.AspNetCore.Razor.TagHelpers.HtmlAttributeValueStyle.DoubleQuotes);
+",
+                csharp,
+                ignoreLineEndingDifferences: true);
+        }
+
+        [Fact]
+        public void WriteSetPreallocatedTagHelperProperty_RendersCorrectly()
+        {
+            // Arrange
+            var extension = new PreallocatedAttributeTargetExtension();
+            var context = new CSharpRenderingContext()
+            {
+                Writer = new CSharpCodeWriter()
+            };
+
+            var descriptor = ITagHelperBoundAttributeDescriptorBuilder
+                .Create("FooTagHelper")
+                .Name("Foo")
+                .TypeName("System.String")
+                .PropertyName("FooProp")
+                .Build();
+
+            var node = new SetPreallocatedTagHelperPropertyIRNode()
+            {
+                AttributeName = descriptor.Name,
+                TagHelperTypeName = "FooTagHelper",
+                VariableName = "_tagHelper1",
+                Descriptor = descriptor,
+            };
+
+            // Act
+            extension.WriteSetPreallocatedTagHelperProperty(context, node);
+
+            // Assert
+            var csharp = context.Writer.Builder.ToString();
+            Assert.Equal(
+@"__FooTagHelper.FooProp = (string)_tagHelper1.Value;
```

Comment:
The newline at the end of this string is significant. And this test is consistent with all our other BasicWriter/TagHelperWriter tests.